core/server/proxy/src/main/java/alluxio/proxy/s3/signature/StringToSignProducer.java

```diff
@@ -0,0 +1,326 @@
+      if (headers.containsKey(header)) {
+        String headerValue = headers.get(header);
+        canonicalHeaders.append(headerValue);
+        canonicalHeaders.append(NEWLINE);
+
+        // Set for testing purpose only to skip date and host validation.
+        validateSignedHeader(schema, header, headerValue);
+      } else {
+        throw new RuntimeException(String.format("%s %s %s", "Header", header,
+                "not present in request but requested to be signed."));
+      }
+    }
+
+    String payloadHash;
+    if (UNSIGNED_PAYLOAD.equals(headers.get(X_AMZ_CONTENT_SHA256))
+            || unsignedPayload) {
+      payloadHash = UNSIGNED_PAYLOAD;
+    } else {
+      payloadHash = headers.get(X_AMZ_CONTENT_SHA256);
+    }
+    StringJoiner canonicalRequestJoiner = new StringJoiner(NEWLINE);
+    canonicalRequestJoiner.add(method)
+            .add(canonicalUri)
+            .add(canonicalQueryStr)
+            .add(canonicalHeaders)
+            .add(signedHeaders)
+            .add(payloadHash);
+    return canonicalRequestJoiner.toString();
+  }
+
+  /**
+   * Returns CanonicalUri.
+   *
+   * @param regex Regular expression to split by
+   * @param uri uri string
+   * @return CanonicalUri
+   */
+  private static String getCanonicalUri(String regex, String uri) {
+    StringJoiner result = new StringJoiner(regex);
+    Pattern p = Pattern.compile(regex);
+    Matcher m = p.matcher(uri);
+    int pos = 0;
+    while (m.find()) {
+      result.add(urlEncode(uri.substring(pos, m.start())));
+      pos = m.end();
+    }
+    result.add(urlEncode(uri.substring(pos)));
+    return result.toString();
+  }
+
+  /**
+   * Encode url.
+   *
+   * Spaces must be encoded as %20. Do not encode spaces as plus signs (+).
+   * Apply the encoding algorithm and replace the plus sign (+) in the encoded
+   * string with %20, the asterisk (*) with %2A, and %7E with the tilde (~).
+   *
+   * @param str string to be encoded
+   * @return the encoded String
+   */
+  private static String urlEncode(String str) {
+    try {
+      return URLEncoder.encode(str, "UTF-8")
+              .replaceAll("\\+", "%20")
+              .replaceAll("%7E", "~");
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String getQueryParamString(Map<String, String> queryMap) {
+    List<String> params = new ArrayList<>(queryMap.keySet());
+
+    // Sort by name, then by value
+    Collections.sort(params, (o1, o2) -> o1.equals(o2)
+            ? queryMap.get(o1).compareTo(queryMap.get(o2))
+            : o1.compareTo(o2));
+
+    StringBuilder result = new StringBuilder();
+    for (String p : params) {
+      if (!p.equals("X-Amz-Signature")) {
+        if (result.length() > 0) {
+          result.append("&");
+        }
+        result.append(urlEncode(p));
+        result.append('=');
+        result.append(urlEncode(queryMap.get(p)));
+      }
+    }
+    return result.toString();
```

Comment:
I don't think query parameters need to be in sorted order, so if you're doing this purely for the string building then I have another suggestion.

You can take a look at this example from `TestCase` on building a string of query parameters using `Joiner` and streams: https://github.com/Alluxio/alluxio/blob/293ea24fb00c57b95a32e8f228c3bd8dd8d9ee47/tests/src/test/java/alluxio/client/rest/TestCase.java#L89-L106

I think you can use a simplified version of that approach because these query parameters all take values.

Reply:
Sorting is required, it affects the signature result.